src/librustc/infer/canonical/canonicalizer.rs

```diff
@@ -385,7 +385,7 @@ impl<'cx, 'gcx, 'tcx> Canonicalizer<'cx, 'gcx, 'tcx> {
             // direct linear search of `var_values`.
             if let Some(idx) = var_values.iter().position(|&k| k == kind) {
                 // `kind` is already present in `var_values`.
-                CanonicalVar::new(idx)
+                DebruijnIndex::new(idx)
```

Comment:
Hmm, this doesn't seem right. To really do this properly, we have to track the number of binders and adjust the debruijn index accordingly.

So, for example, if we are canonicalizing something like this:

```rust
for<'a> fn(&'a _)
```

then there is one binder in scope (from the `fn`) and hence I would expect the debruijn index to be adusted.

OTOH, I suppose that — at present — these canonical variables effectively occupy a "separate universe" from late-bound regions and things, and hence it doesn't make sense to adjust for the number of binders in scope. I think that's why this code all works fine. 

Hmm, so maybe the PR is right "as is", at least as a stepping stone to the final intended setup.

Reply:
Ok. Let me know when you have made a decision. I suppose a gradual but not strictly correct approach wouldn't be too bad...

For a correct solution I would probably need a bit of mentoring, since I am not sure if I understand the problem completely.